scope.go

```diff
@@ -0,0 +1,260 @@
+// there are no implementations.
+type ScopeOption interface {
+	applyScopeOption(*scopeOptions)
+}
+
+type scopeOptions struct {
+}
+
+// Scope is a scoped DAG of types and their dependencies.
+// A Scope may also have one or more child Scopes that inherit
+// from it.
+type Scope struct {
+	name string
+	// Mapping from key to all the constructor node that can provide a value for that
+	// key.
+	providers map[key][]*constructorNode
+
+	// constructorNodes provided directly to this Scope. i.e. it does not include
+	// any nodes that were provided to the parent Scope this inherited from.
+	nodes []*constructorNode
+
+	// Values that have already been generated directly in the Scope.
+	values map[key]reflect.Value
+
+	// Values groups that have already been generated directly in the Scope.
+	groups map[key][]reflect.Value
+
+	// Source of randomness.
+	rand *rand.Rand
+
+	// Flag indicating whether the graph has been checked for cycles.
+	isVerifiedAcyclic bool
+
+	// Defer acyclic check on provide until Invoke.
+	deferAcyclicVerification bool
+
+	// invokerFn calls a function with arguments provided to Provide or Invoke.
+	invokerFn invokerFn
+
+	// graph of this Scope. Note that this holds the dependency graph of all the
+	// nodes that affect this Scope, not just the ones provided directly to this Scope.
+	gh *graphHolder
+
+	// Parent of this Scope.
+	parentScope *Scope
+
+	// All the child scopes of this Scope.
+	childScopes []*Scope
+}
+
+// Scope creates a new Scope with the given name and options from current Scope.
+// Any constructors that the current Scope knows about, as well as any modifications
+// made to it in the future will be propagated to the child scope.
+// However, no modifications made to the child scope being created will be propagated
+// to the parent Scope.
+func (s *Scope) Scope(name string, opts ...ScopeOption) *Scope {
+	child := &Scope{
+		name:        name,
+		parentScope: s,
+		providers:   make(map[key][]*constructorNode, len(s.providers)),
+		values:      make(map[key]reflect.Value, len(s.values)),
+		groups:      make(map[key][]reflect.Value, len(s.groups)),
+		invokerFn:   s.invokerFn,
+	}
+
+	// child should hold a separate graph holder
+	child.gh = &graphHolder{
+		s:     child,
+		snap:  -1,
+		nodes: make([]*graphNode, len(s.gh.nodes)),
+	}
+
+	for i, graphNode := range s.gh.nodes {
+		child.gh.nodes[i] = graphNode
+	}
+
+	s.childScopes = append(s.childScopes, child)
+	return child
+}
+
+// getScopesUntilRoot creates a list of Scopes
+// have to traverse through until the current node.
+func (s *Scope) getScopesUntilRoot() []*Scope {
+	if s.parentScope != nil {
+		return append(s.parentScope.getScopesUntilRoot(), s)
+	}
+	return []*Scope{s}
+}
+
+func (s *Scope) getAllLeafScopes() []*Scope {
+	allScopes := []*Scope{s}
+	if len(s.childScopes) > 0 {
+		for _, cs := range s.childScopes {
+			allScopes = append(allScopes, cs.getAllLeafScopes()...)
+		}
+	}
+	return allScopes
```

Comment:
FWIW this will do a lot of allocs because every call will allocate that slice only for it to be thrown away and copied into the parent's slice. An alternative API here is:

```
func (s *Scope) appendLeafScopes(dest []*Scope) []*Scope {
  dest = append(dest, s)
  for _, c := range s.childScopes {
    dest = c.appendLeafScopes(dest)
  }
  return dest
}
```

Usage: `s.appendLeafScopes(nil)` (because nil is a valid slice).
This will allocate a single slice in total.